### PR TITLE
Enhancement: Raise error for filter expressions which do not evaluate to `bool`

### PR DESCRIPTION
**Pull request by tedil**
This PR addresses #154 by adding a check on the result of the filter expression. If it's `bool`, all is well, otherwise, print a warning but continue, because someone might want to use a truth-y value as a filter.

### CODE REVIEW

**Comment by tedil** [comment]
As an alternative, we could also simply force booleans via `keep = bool(self._func())`.

**Comment by dlaehnemann** [comment]
I think a warning is a good idea, here. It should make it easier for users to catch this as unwanted behaviour and allow to use this behaviour if users need it.

**Comment by christopher-schroeder** [comment]
I agree. Maybe an option to disable specific warnings? That should satisfy everyone

**Comment by dlaehnemann** [comment]
And maybe also an addition to the `README.md` docs along the lines that @ptrebert suggested:
https://github.com/vembrane/vembrane/issues/154#issuecomment-1655621757

I would link out to the Truth Value Testing docs of Python that @tedil mentioned and inlcude the statement that a non-empty list will evaluate to `True`.

**Comment by tedil** [comment]
I also thought about whether it is possible to derive from the expression itself whether its result is of type bool or not, but since Python is not a statically typed language, it's probably not particularly easy to get right. 
Instead, one could peek at the first record, apply the filter expression and check that result (and then issue a warning/exit early, whatever).
Though I can imagine that, potentially for some VCF file and expression combinations, the expression may result in different types for different records (such as `0 if INFO['foo'] else "bar"`, which does not make much sense, though.).

Perhaps raising an error instead of a warning would be indeed better, as Peter suggested.
If anyone really wants to rely on truthy values, wrapping the expression in `bool(…)` is always possible and way more explicit.

File: README.md [hunk 1]
```diff
@@ -42,7 +42,9 @@ options:
 
 
 ### Filter expression
-The filter expression can be any valid python expression that evaluates to `bool`. However, functions and symbols available have been restricted to the following:
+The filter expression can be any valid python expression that evaluates to a value of *type* `bool`.
```
Comment (dlaehnemann) [inline README.md:45]:
```suggestion
The filter expression can be any valid python expression that evaluates to a value of *type* `bool`.
If you want to use truthy values, you need to wrap the expression in `bool()`, or aggregate multiple values via `any()` or `all()`.
```

**Review by dlaehnemann — APPROVED** [review]
I like this approach, it seems like the most sensible. Two things:

1. I would include the new `NonBoolTypeError` help message's info in the main docs as well, as I think this is helpful information for any user and might even make vembrane's behaviour clearer.
2. We should probably bump the major version, as this changes how certain expressions are handled.

**Comment by dlaehnemann** [comment]
That almost sounds like a new Application Note, right? :D

**Pull request merged**